src/main/java/jenkins/scm/api/SCMFile.java

```diff
@@ -41,73 +44,237 @@
  * @author Kohsuke Kawaguchi
  */
 public abstract class SCMFile {
+
+    /**
+     * The parent {@link SCMFile} or {@code null} for the root.
+     */
+    @CheckForNull
+    private final SCMFile parent;
+    /**
+     * The name of this {@link SCMFile}
+     */
+    private final String name;
+    /**
+     * Cache of the file type information, to allow repeated calls to minimize the number of network round trips.
+     * We cache the type information because too many people think that
+     * {@code File f; if (f.exists() && f.isFile()) ...} is a good code pattern, ignoring the fact that
+     * {@link File#isFile()} includes and existence check. We provide {@link #getType()} to avoid the need for
+     * such calling styles, but when people are replicating local code it can be easier to keep the logic
+     * the same as the code you are replicating - which is why we cache this here.
+     */
+    private Type type;
+
+    /**
+     * Constructor for the root entry.
+     *
+     * @since 2.0
+     */
+    protected SCMFile() {
+        this.parent = null;
+        this.name = "";
+    }
+
+    /**
+     * Constructor for any entry that is not the root.
+     *
+     * @param parent the parent reference or {@code null} if this is the root object.
+     * @param name   the name of this entry.
+     * @since 2.0
+     */
+    protected SCMFile(@NonNull SCMFile parent, String name) {
+        this.parent = parent;
+        this.name = name;
+    }
+
     /**
      * Gets the file name of this file without any path portion, such as just "foo.txt"
      * <p>This method is the equivalent of {@link File#getName()}.</p>
      *
      * @return the file name of this file without any path portion.
      */
     @NonNull
-    public abstract String getName();
+    public final String getName() {
+        return name;
+    }
 
     /**
-     * Gets a child/descendant path relative from this object.
+     * Gets the file name including the path portion, such as "foo/bar/manchu.txt". Will never end in {@code /}.
      *
-     * @param path Relative path of the child to return.
-     * @return null if there's no file/directory at the path represented by it.
-     * @throws IOException if an error occurs while performing the operation.
+     * @return the pathname of this file.
+     * @since 2.0
+     */
+    @NonNull
+    public String getPath() {
+        if (parent == null) {
+            // root node
+            return "";
```

Comment:
You seem to be including the name below so shouldn't this be returning `getName()`?